### PR TITLE
Refactor measurement flow and API.

### PR DESCRIPTION
**Pull request by NullHypothesis**
This PR makes a number of significant changes:

- Make the zerotrace object safe to use concurrently.
- Parallelize the sending of trace packets.
- Refactor the internals to be event-driven by using `for`/`select` loops.
- Adapt the package's API.
- Delete unnecessary code.

### CODE REVIEW

File: ipids.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
+package zerotrace
+
+import (
+	"errors"
+	"math"
+	"math/rand"
```
[the remark this answers is not in the document]
Comment (NullHypothesis) [reply ipids.go:6]: False positive. Our use of `rand` has no security implications.

File: ipids.go [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+	s.Lock()
+	defer s.Unlock()
+
+	if len(s.ipids) == math.MaxUint16 {
+		return 0, errNoMoreIds
+	}
+
+	// Start at a random index and look for available IP IDs.
+	start := uint16(rand.Intn(math.MaxUint16))
+	for id := start + 1; id != start; id++ {
```
Comment (rillian) [inline ipids.go:47]: Does this wrap around to 0, or run off the end of the map? `rand.Intn` returns a value from the half-open interval [0, math.MaxUint16] so the maximum value possible is 65534. Adding 1 starts the search at 65535, which is the last possible key, leaving only one slot to check. The `id != start` termination check suggests it's wrapping, but wanted to confirm?

File: ipids_test.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+func BenchmarkBorrow(b *testing.B) {
+	var (
+		err error
+		p   *ipIdPool
+	)
+
+	for i := 0; i < b.N; i++ {
+		p = newIpIdPool()
+		for j := 0; j < math.MaxUint16; j++ {
```
Comment (rillian) [inline ipids_test.go]: `<=` ?

File: util.go [hunk 4]
```diff
@@ -34,21 +37,23 @@ func extractTTL(ipPkt []byte) (uint8, error) {
 		return 0, err
 	}
 
-	return uint8(ipPkt[8]), nil
+	return uint16(ipPkt[4])<<8 | uint16(ipPkt[5]), nil
```
Comment (rillian) [inline util.go]: Would this be easier to read as

```go
// IP Identification field is in bytes 4-5 of the header.
return binary.BigEndian.Uint16(ipPkt[4:])
```
(from encoding/binary in the standard library)?

File: util_test.go [hunk 5]
```diff
@@ -43,22 +43,3 @@ func TestExtractIPID(t *testing.T) {
 		t.Fatalf("Expected IP ID %d but got %d.", expectedIPID, ipID)
 	}
 }
-
-func TestExtractTTL(t *testing.T) {
```
Comment (rillian) [inline util_test.go:47]: No `TestExtractIPID`?

File: zerotrace.go [hunk 6]
```diff
@@ -15,223 +15,221 @@ import (
...
-			if ipLayer == nil || icmpLayer == nil {
-				continue
-			}
-
-			// If it is an ICMP packet, check if it is the ICMP TTL
-			// exceeded one we are looking for
-			respPkt, err := z.extractRcvdPkt(pkt)
+		case <-ticker.C:
+			z.ipids.releaseUnanswered()
+			l.Printf("Released un-answered IP IDs; %d left.", z.ipids.size())
```
Comment (rillian) [inline zerotrace.go]: Possibly redundant with the log message inside `releaseUnanswered`?

**Review by rillian — APPROVED** [review]
Looks reasonable to me. Some minor questions below.

File: ipids.go (hunk 2 above)
Comment (NullHypothesis) [reply ipids.go:47]: Yes, it wraps around. Documented in 45622cf2b522c1cded57da9c88e3c1018180f909.

File: ipids_test.go (hunk 3 above)
Comment (NullHypothesis) [reply ipids_test.go]: Good catch! Fixed in a3b04a166696ca0243feb7c74c8d55d4df23d84e.

File: util.go (hunk 4 above)
Comment (NullHypothesis) [reply util.go]: Agreed. Fixed in aac5e581c3ef92c4f3938d634a8605e8533255d8.

File: util_test.go (hunk 5 above)
Comment (NullHypothesis) [reply util_test.go:47]: Not sure I follow. `TestExtractIPID` is [here](https://github.com/brave/zerotrace/blob/aac5e581c3ef92c4f3938d634a8605e8533255d8/util_test.go#L28-L45).

File: zerotrace.go (hunk 6 above)
Comment (NullHypothesis) [reply zerotrace.go]: Yes, fixed in c6224beeecb951ba338406dfeb9877631b665214.

File: util_test.go (hunk 5 above)
Comment (rillian) [reply util_test.go:47]: Sorry, missed it in the diff.

File: ipids.go (hunk 2 above)
Comment (rillian) [reply ipids.go:47]: Great, thanks.

**Pull request merged**